### PR TITLE
Introduce cached time source

### PR DESCRIPTION
**Pull request by oschaaf**
This helps achieve a consistent view of “now” across the sequencer and its associated rate limiter(s) and termination predicate(s). This should help avoiding subtle edge cases caused by small differences when these components all query the clock independently (and possibly TOCTOU).

------

Prerequisite:

- [x] Land #285, from which this change was split out.
- [x] Merge master

### CODE REVIEW

**Comment by mum4k** [comment]
@dubious90 please review and assign to me when ready.

@htuch, this feels like a change where your feedback would be valuable.

**Comment by mum4k** [comment]
@oschaaf please merge master back here after phases were merged to ease the review.

**Comment by oschaaf** [comment]
> @oschaaf please merge master back here after phases were merged to ease the review.

@mum4k done!

File: source/common/sequencer_impl.cc [hunk 1]
```diff
@@ -81,6 +81,9 @@ void SequencerImpl::updateStartBlockingTimeIfNeeded() {
 
 void SequencerImpl::run(bool from_periodic_timer) {
   ASSERT(running_);
+  // Update time so that our time source will yield an up-to-date monotonic time sample to work with
+  // during this function call.
+  dispatcher_.updateApproximateMonotonicTime();
```
Comment (pemor) [inline source/common/sequencer_impl.cc]: if I understand this correctly, then real monotonicTime() is called in updateApproximateMonotonicTime, which means no optimization gain for this specific call. 
Can you please call out where the cached_time_source..monotonicTime() would be saving time.

File: source/client/client_worker_impl.h [hunk 2]
```diff
@@ -48,6 +48,7 @@ class ClientWorkerImpl : public WorkerImpl, virtual public ClientWorker {
 private:
   void simpleWarmup();
 
+  std::unique_ptr<Envoy::TimeSource> cached_time_source_;
```
[the remark this answers is not in the document]
Comment (oschaaf) [reply source/client/client_worker_impl.h]: I opted for a smart pointer here so we can force the implementation to go through the (abstract) `Envoy::TimeSource` interface. E.g. maybe we'll some day want an escape latch option to use the `regular` time source` and then it's just a matter of initializing this differently?

File: source/common/sequencer_impl.cc (hunk 1 above)
Comment (oschaaf) [reply source/common/sequencer_impl.cc]: I added a more elaborate comment in 823c15eba8d6261edb832840479e1c69bdd1d7e8, does that look helpful?

File: source/client/client_worker_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (pemor) [reply source/client/client_worker_impl.h]: make sense, sorry I missed it.

File: source/common/sequencer_impl.cc (hunk 1 above)
Comment (pemor) [reply source/common/sequencer_impl.cc]: Very helpful. Thanks.

**Pull request merged**